Tw 789 sequencer architecture and transaction flow deep dive (#3471)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/sequencer-transaction-flow.mdx

---
title: Sequencer architecture and transaction flow
description: 'A deep dive into how the Sequencer processes transactions end-to-end: the transaction queue, ordering, block creation timing, and the sequencer feed.'
author: Gowtham118
sme: Jason Wan
user_story: As a chain operator or integration partner, I need to understand how the Sequencer processes transactions end-to-end.
content_type: concept
---

> This deep dive follows a transaction through a single <a data-quicklook-from="sequencer">Sequencer</a> instance: how it arrives, how it waits in the transaction queue, how blocks get created, and how the result reaches the rest of the network. It is aimed at chain operators and integration partners who need to reason about queueing, timeouts, and block timing.

Reviewer: ```suggestion
This deep dive follows a <a data-quicklook-from="transaction">transaction</a> through a single <a data-quicklook-from="sequencer">Sequencer</a> instance: how it arrives, how it waits in the transaction queue, how blocks get created, and how the result reaches the rest of the network. It is aimed at chain operators and integration partners who need to reason about queueing, timeouts, and block timing.
```

Two companion pages cover the surrounding context, and this page assumes you have read them:

- [The Sequencer and censorship resistance](/how-arbitrum-works/deep-dives/sequencer.mdx) explains the Sequencer's role, the real-time feed, batch posting, and finality.
- [Transaction lifecycle](/how-arbitrum-works/deep-dives/transaction-lifecycle.mdx) explains the different ways to submit a transaction, including bypassing the Sequencer entirely.

:::note Scope of this page

> This page describes the internals of a single Sequencer instance. For the behavior of <a data-quicklook-from="arbitrum-one">Arbitrum One</a>'s public sequencer endpoint (latency expectations, retries, and fallback patterns), see [RPC endpoints and providers](/arbitrum-essentials/reference/node-providers.mdx). For running multiple redundant Sequencers with Redis-based coordination, see [How to set up a high-availability sequencer](/launch-arbitrum-chain/run-a-node/high-availability-sequencer.mdx) and [How to run a Sequencer Coordinator Manager](/run-arbitrum-node/sequencer/03-run-sequencer-coordination-manager.mdx).

Reviewer: ```suggestion
This page describes the internals of a single Sequencer instance. For the behavior of <a data-quicklook-from="arbitrum-one">Arbitrum One</a>'s public sequencer endpoint (latency expectations, retries, and fallback patterns), see [RPC endpoints and providers](/arbitrum-essentials/reference/node-providers.mdx). For running multiple redundant Sequencers with Redis-based coordination, see [How to set up a high-availability sequencer](/launch-arbitrum-chain/run-a-node/high-availability-sequencer.mdx) and [How to run a Sequencer Coordination Manager](/run-arbitrum-node/sequencer/03-run-sequencer-coordination-manager.mdx).
```

:::

All code references below point to [Nitro `v3.11.0`](https://github.com/OffchainLabs/nitro/tree/a618155919315241665356fe60f3cd00d66d5e46).

## Transaction flow at a glance

> <ImageZoom src="/img/haw-sequencer-transaction-flow.svg" alt="Transaction flow from a user through an RPC node into the Sequencer's bounded transaction queue, FIFO into block creation, then out to the sequencer feed and, via the batch poster, to the sequencer inbox on the parent chain where validators verify it" className="img-600px" />

Reviewer: ```suggestion
<ImageZoom src="/img/haw-sequencer-transaction-flow.svg" alt="Transaction flow from a user through an RPC node into the Sequencer's bounded transaction queue, FIFO into block creation, then out to the sequencer feed and, via the batch poster, to the sequencer inbox on the parent chain where validators verify it" className="img-800px" />
```

1. A user submits a signed transaction to any RPC node on the chain.
2. The RPC node does not execute or pool the transaction; it forwards it to the Sequencer.
3. The Sequencer places the transaction in a bounded, in-memory queue.
4. The block creation loop drains the queue in FIFO order and executes transactions one at a time to build a block.
5. The new block is published on the <a data-quicklook-from="sequencer-feed">Sequencer Feed</a> for real-time consumers, and the <a data-quicklook-from="batch-poster">batch poster</a> later posts the compressed sequence to the <a data-quicklook-from="sequencer-inbox">sequencer inbox</a> on the <a data-quicklook-from="parent-chain">parent chain</a>.
6. <a data-quicklook-from="validator">Validators</a> re-execute the sequence posted on the parent chain to verify and assert the chain's state.

> :::note Validators do not accept user transactions

Reviewer: Great point

> Validators only read the transaction sequence from the parent chain and verify the resulting state. They have no transaction queue and no way to accept, order, or process a user transaction directly. The only ingestion point for user transactions is the Sequencer, and RPC nodes exist to forward transactions to it. (The one exception, submitting through the <a data-quicklook-from="delayed-inbox">delayed inbox</a> on the parent chain, is covered in [Transaction lifecycle](/how-arbitrum-works/deep-dives/transaction-lifecycle.mdx).)

Reviewer 2: ```suggestion
Validators only read the ordered transactions from the parent chain and re-execute them locally to compute the chain's state. They have no transaction queue and no way to accept, order, or process a user transaction directly. The only ingestion point for user transactions is the Sequencer, and RPC nodes exist to forward transactions to it. (The one exception, submitting through the <a data-quicklook-from="delayed-inbox">delayed inbox</a> on the parent chain, is covered in [Transaction lifecycle](/how-arbitrum-works/deep-dives/transaction-lifecycle.mdx).)
```
The original content may mislead users think sequencer will also post the resulting state

Reviewer 2: Another suggestion: We mentioned fifo several times here, should we add timeboost will behavior different here?

:::

## How transactions reach the Sequencer

> Only one node on the chain runs with sequencing enabled. Every other node runs a **forwarder**: when it receives a transaction over RPC, it immediately relays the raw transaction to the Sequencer's endpoint instead of processing it locally ([`forwarder.go`](https://github.com/OffchainLabs/nitro/blob/a618155919315241665356fe60f3cd00d66d5e46/execution/gethexec/forwarder.go#L126)). The target is configured with `--execution.forwarding-target`; for known chains, Nitro fills it in automatically from the chain's configuration when the node is not a sequencer. A non-sequencer node with no forwarding target (explicitly set to `"null"`) simply drops incoming transactions.

Reviewer 2: ```suggestion
Only one node on the chain actively sequences transactions at any given time. Every other node runs a **forwarder**: when it receives a transaction over RPC, it immediately relays the raw transaction to the Sequencer's endpoint instead of processing it locally ([`forwarder.go`](https://github.com/OffchainLabs/nitro/blob/a618155919315241665356fe60f3cd00d66d5e46/execution/gethexec/forwarder.go#L126)). The target is configured with `--execution.forwarding-target`; for known chains, Nitro fills it in automatically from the chain's configuration when the node is not a sequencer. A non-sequencer node with no forwarding target (explicitly set to `"null"`) simply drops incoming transactions.
```
we need to consider seq coordinator system as well
